### PR TITLE
fix: macOS e2e local tests fails on M3 silicon processors

### PR DESCRIPTION
**Pull request by falcucci**
Some crucial dependencies were missing to make the cardano-cli and cardano-node to work as it should in a standalone environment.

after-all, this aims to bring more compatibility for a general local troubleshooting due to continuous errors the script has shown to start the nodes, as you can see as follows:

```bash
cargo build --release -p mithril-aggregator -p mithril-signer -p mithril-client-cli -p mithril-relay && cargo run -p mithril-end-to-end -- -vvvvvvvvvv --bin-directory target/release/ --work-directory=$(PWD)/artifacts --devnet-scripts-directory=./mithril-test-lab/mithril-devnet --cardano-node-version 8.7.3
    Finished `release` profile [optimized] target(s) in 0.20s
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/mithril-end-to-end -vvvvvvvvvv --bin-directory target/release/ --work-directory=/Users/falcucci/txpipe/mithril/artifacts --devnet-scripts-directory=./mithril-test-lab/mithril-devnet --cardano-node-version 8.7.3`
Jun 18 23:05:32.395 INFO Bootstrapping the Devnet, script: /Users/falcucci/txpipe/mithril/mithril-test-lab/mithril-devnet/devnet-mkfiles.sh
tar: ./bin/cardano-cli: Not found in archive
tar: ./bin/cardano-node: Not found in archive
tar: Error exit delayed from previous errors.
dyld[38816]: Library not loaded: @executable_path/libsodium.23.dylib
  Referenced from: <no uuid> /Users/falcucci/txpipe/mithril/artifacts/devnet/cardano-cli
  Reason: tried: '/Users/falcucci/txpipe/mithril/artifacts/devnet/libsodium.23.dylib' (no such file), '/usr/local/lib/libsodium.23.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/lib/libsodium.23.dylib' (no such file, not in dyld cache)
/Users/falcucci/txpipe/mithril/mithril-test-lab/mithril-devnet/mkfiles/mkfiles-cardano.sh: line 81: 38816 Abort trap: 6           $CARDANO_CLI byron genesis genesis --protocol-magic ${NETWORK_MAGIC} --start-time "${START_TIME}" --k ${SECURITY_PARAM} --n-poor-addresses 0 --n-delegate-addresses ${NUM_SPO_NODES} --total-balance ${INIT_SUPPLY} --delegate-share 1 --avvm-entry-count 0 --avvm-entry-balance 0 --protocol-parameters-file "${ARTIFACTS_DIR_TEMP}/byron.genesis.spec.json" --genesis-output-dir "${ARTIFACTS_DIR_TEMP}/byron-gen-command"
Jun 18 23:05:35.776 INFO Starting the Devnet, script: /Users/falcucci/txpipe/mithril/artifacts/devnet/start-cardano.sh
Error: Failed to start the devnet

Caused by:
    No such file or directory (os error 2)
```

### CODE REVIEW

**Comment by dlachaume** [comment]
Hi @falcucci!

Did you try running the E2E without specifying the version option for the cardano-node? The version `8.7.3` is a bit old now.

On my local environment (MacOS M2), I'm running the E2E with this command without any issue:
`cargo run -p mithril-end-to-end -- -vvv --bin-directory target/release/ --devnet-scripts-directory=./mithril-test-lab/mithril-devnet --work-directory=/tmp/mithril_end_to_end`

**Comment by falcucci** [comment]
@dlachaume hey, it fails specifying any node version and even without it.

the cardano-node version 8.7.3 is part of the e2e tests scenarios that we have ran currently in the CI steps as the same as
- 8.9.0
- 8.11.0
- 8.1.2

So it should also be replicable.

Anyway this isn't a node-version issue specifically but the setup of the libraries itself.

```bash
cargo run -p mithril-end-to-end -- -vvv --bin-directory target/release/ --devnet-scripts-directory=./mithril-test-lab/mithril-devnet --work-directory=/tmp/mithril_end_to_end
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.34s
     Running `target/debug/mithril-end-to-end -vvv --bin-directory target/release/ --devnet-scripts-directory=./mithril-test-lab/mithril-devnet --work-directory=/tmp/mithril_end_to_end`
Jun 19 10:31:56.835 INFO Bootstrapping the Devnet, script: /Users/falcucci/txpipe/mithril/mithril-test-lab/mithril-devnet/devnet-mkfiles.sh
dyld[51555]: Library not loaded: @executable_path/libsodium.23.dylib
  Referenced from: <no uuid> /private/tmp/mithril_end_to_end/devnet/cardano-cli
  Reason: tried: '/private/tmp/mithril_end_to_end/devnet/libsodium.23.dylib' (no such file), '/usr/local/lib/libsodium.23.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/usr/lib/libsodium.23.dylib' (no such file, not in dyld cache)
/Users/falcucci/txpipe/mithril/mithril-test-lab/mithril-devnet/mkfiles/mkfiles-cardano.sh: line 81: 51555 Abort trap: 6           $CARDANO_CLI byron genesis genesis --protocol-magic ${NETWORK_MAGIC} --start-time "${START_TIME}" --k ${SECURITY_PARAM} --n-poor-addresses 0 --n-delegate-addresses ${NUM_SPO_NODES} --total-balance ${INIT_SUPPLY} --delegate-share 1 --avvm-entry-count 0 --avvm-entry-balance 0 --protocol-parameters-file "${ARTIFACTS_DIR_TEMP}/byron.genesis.spec.json" --genesis-output-dir "${ARTIFACTS_DIR_TEMP}/byron-gen-command"
Jun 19 10:31:59.822 INFO Starting the Devnet, script: /private/tmp/mithril_end_to_end/devnet/start-cardano.sh
Error: Failed to start the devnet

Caused by:
    No such file or directory (os error 2)
```

**Comment by falcucci** [comment]
@Alenar that makes total sense, moving forward to an isolated `/bin` is a better approach.

I can handle it when I have some free time soon 🙂

File: mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh [hunk 1]
```diff
@@ -65,20 +65,21 @@ if ! mkdir -p "${ARTIFACTS_DIR}"; then
...
+echo ">> Creating artifacts directory..."
 # Download cardano-cli & cardano-node if enabled (default: yes)
 if [[ "$SKIP_CARDANO_BIN_DOWNLOAD" != "true" ]]; then
   echo ">> Downloading cardano-cli & cardano-node..."
-  curl -sL ${CARDANO_BINARY_URL} --output cardano-bin.tar.gz
-  tar xzf cardano-bin.tar.gz ./bin/cardano-cli ./bin/cardano-node && mv ./bin/cardano-{cli,node} . && rm -rf ./bin  || tar xzf cardano-bin.tar.gz ./cardano-cli ./cardano-node
+  curl -sL "${CARDANO_BINARY_URL}" --output cardano-bin.tar.gz
+  echo ">> Extracting cardano-cli & cardano-node..."
+  tar xzf cardano-bin.tar.gz ./bin || (mkdir -p ./bin && tar --strip-components=1 -C ./bin -xzf cardano-bin.tar.gz)
   rm -f cardano-bin.tar.gz
```
Comment (Alenar) [inline mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh:75]: You could directly decompress to the artifact directory avoiding the cp/rm that come afterward :).
```suggestion
  tar xzf cardano-bin.tar.gz ./bin -C "${ARTIFACTS_DIR}/" || (mkdir -p "${ARTIFACTS_DIR}/bin" && tar --strip-components=1 -C "${ARTIFACTS_DIR}/bin" -xzf cardano-bin.tar.gz)
  rm -f cardano-bin.tar.gz
```

File: mithril-test-lab/mithril-devnet/.gitignore
Comment (Alenar) [inline mithril-test-lab/mithril-devnet/.gitignore:1]: Are those additions still needed ? With your current changes nothing new should be decompress in the `mithril-devnet` directory isn't it ?

File: mithril-test-lab/mithril-devnet/devnet-mkfiles.sh [hunk 2]
```diff
@@ -6,6 +6,7 @@ set -e
 if [[ -n $DEBUG ]]; then
     set -x
 fi
+set -x
```
Comment (Alenar) [inline mithril-test-lab/mithril-devnet/devnet-mkfiles.sh:9]: Can you revert this change ? The block just above already allow to set it by setting a "DEBUG" env var to any value before calling the script :).

**Review by Alenar — COMMENTED** [review]
LGTM ! Just a few issue remaining :rocket: .

File: mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh (hunk 1 above)
Comment (falcucci) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh]: thank you @Alenar

**Review by dlachaume — APPROVED** [review]
Thanks for your contribution @falcucci!
Just the last comment from @Alenar to take into account and an update of your branch from the main branch.
Otherwise, LGTM  🚀

File: mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh [hunk 3]
```diff
@@ -65,20 +65,21 @@ if ! mkdir -p "${ARTIFACTS_DIR}"; then
...
+  echo ">> Extracting cardano-cli & cardano-node..."
+  tar xzf cardano-bin.tar.gz ./bin -C "${ARTIFACTS_DIR}/" || (mkdir -p "${ARTIFACTS_DIR}/bin" && tar --strip-components=1 -C "${ARTIFACTS_DIR}/bin" -xzf cardano-bin.tar.gz)
   rm -f cardano-bin.tar.gz
 fi
 
-# And copy cardano-cli & cardano-node
-cp cardano-cli ${ARTIFACTS_DIR}/cardano-cli
-cp cardano-node ${ARTIFACTS_DIR}/cardano-node
+cp -r ./bin "${ARTIFACTS_DIR}/" 2>/dev/null || true
+rm -rf ./bin
```
[the remark this answers is not in the document]
Comment (falcucci) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh:79]: hey @Alenar 

actually this specific suggestion is crashing the tests now by using

```bash
tar xzf cardano-bin.tar.gz ./bin -C "${ARTIFACTS_DIR}/" || (mkdir -p "${ARTIFACTS_DIR}/bin" && tar --strip-components=1 -C "${ARTIFACTS_DIR}/bin" -xzf cardano-bin.tar.gz)
```

and removing the decompression part:

```bash
cp -r ./bin "${ARTIFACTS_DIR}/" 2>/dev/null || true
rm -rf ./bin
```

here you have some logs:

```bash
cargo build --release -p mithril-aggregator -p mithril-signer -p mithril-client-cli -p mithril-relay && cargo run -p mithril-end-to-end -- -vvvvvvvvvv --bin-directory target/release/ --work-directory=$(PWD)/artifacts --devnet-scripts-directory=./mithril-test-lab/mithril-devnet --cardano-node-version 9.1.0
    Finished `release` profile [optimized] target(s) in 0.20s
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.16s
     Running `target/debug/mithril-end-to-end -vvvvvvvvvv --bin-directory target/release/ --work-directory=/Users/falcucci/txpipe/mithril/artifacts --devnet-scripts-directory=./mithril-test-lab/mithril-devnet --cardano-node-version 9.1.0`
Aug 20 17:20:56.065 INFO Bootstrapping the Devnet, script: /Users/falcucci/txpipe/mithril/mithril-test-lab/mithril-devnet/devnet-mkfiles.sh
tar: -C: Not found in archive
tar: /Users/falcucci/txpipe/mithril/artifacts/devnet: Not found in archive
tar: Error exit delayed from previous errors.
/Users/falcucci/txpipe/mithril/mithril-test-lab/mithril-devnet/mkfiles/mkfiles-cardano.sh: line 72: ./bin/cardano-cli: No such file or directory
Aug 20 17:20:59.689 INFO Starting the Devnet, script: /Users/falcucci/txpipe/mithril/artifacts/devnet/start-cardano.sh
Error: Failed to start the devnet

Caused by:
    No such file or directory (os error 2)
```

[the remark this answers is not in the document]
Comment (falcucci) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh:79]: the previous approach was working just fine,

actually I would say it's much clearer in terms of readability the previous way isolating those steps by:

```bash
tar xzf cardano-bin.tar.gz ./bin || (mkdir -p ./bin && tar --strip-components=1 -C ./bin -xzf cardano-bin.tar.gz)
```

and then using 

```bash
cp -r ./bin "${ARTIFACTS_DIR}/" 2>/dev/null || true
rm -rf ./bin
```

[the remark this answers is not in the document]
Comment (falcucci) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh:79]: but this is a take it or leave it, I can also apply your suggestion once it stops crashing and has no-warnings.

[the remark this answers is not in the document]
Comment (Alenar) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh:79]: I checked it on my side and the error is due to argument order on the first tar command, the `-C  "${ARTIFACTS_DIR}"` must come before the `./bin` since it only apply to the arguments after it (my bad :bow: ).

So the tar line should be:
```bash
tar xzf cardano-bin.tar.gz -C "${ARTIFACTS_DIR}/" ./bin || (mkdir -p "${ARTIFACTS_DIR}/bin" && tar --strip-components=1 -C "${ARTIFACTS_DIR}/bin" -xzf cardano-bin.tar.gz)
```

This should work like that, if there's still errors even with this fix you can revert to your previous proposition so we can merge.

[the remark this answers is not in the document]
Comment (falcucci) [reply mithril-test-lab/mithril-devnet/mkfiles/mkfiles-init.sh]: thank you @Alenar 

done.

**Review by Alenar — APPROVED** [review]

**Review by jpraynaud — APPROVED** [review]
LGTM :+1: 

Thanks for the contribution @falcucci! 

Can you bump the patch version of the devnet (`VERSION` file in the `mithril-devnet` folder) so that we can merge the PR?
